tools/duplicate_function_eliminator.py

```diff
@@ -0,0 +1,355 @@
+      pool = multiprocessing.Pool(processes=cores)
+      filenames = pool.map(run_on_chunk, commands, chunksize=1)
+    else:
+      # We can't parallize, but still break into chunks to avoid uglify/node memory issues
+      if len(chunks) > 1 and DEBUG: print >> sys.stderr, 'splitting up js optimization into %d chunks' % (len(chunks))
+      filenames = [run_on_chunk(command) for command in commands]
+  else:
+    filenames = []
+
+  json_files = []
+
+  # We're going to be coalescing the files back at the end
+  # Just replace the file list with the ones provided in
+  # the command list - and save off the generated Json
+  if gen_hash_info:
+    json_files = filenames[:]
+    filenames = old_filenames[:]
+
+  for filename in filenames: temp_files.note(filename)
+
+  filename += '.jo.js'
+  f = open(filename, 'w')
+  f.write(pre);
+  pre = None
+
+  # sort functions by size, to make diffing easier and to improve aot times
+  funcses = []
+  for out_file in filenames:
+    funcses.append(split_funcs(open(out_file).read(), False))
+  funcs = [item for sublist in funcses for item in sublist]
+  funcses = None
+  def sorter(x, y):
+    diff = len(y[1]) - len(x[1])
+    if diff != 0: return diff
+    if x[0] < y[0]: return 1
+    elif x[0] > y[0]: return -1
+    return 0
+  if not os.environ.get('EMCC_NO_OPT_SORT'):
+    funcs.sort(sorter)
+
+  for func in funcs:
+    f.write(func[1])
+  funcs = None
+
+  f.write('\n')
+  f.write(post);
+  # No need to write suffix: if there was one, it is inside post which exists when suffix is there
+  f.write('\n')
+
+  if gen_hash_info and len(json_files) > 0:
+    write_equivalent_fn_hash_to_file(f, json_files, passed_in_filename)
+  f.close()
+
+  return filename
+
+def save_temp_file(file_to_process):
+  if os.environ.get('EMSCRIPTEN_SAVE_TEMP_FILES') and os.environ.get('EMSCRIPTEN_TEMP_FILES_DIR'):
+    destinationFile = file_to_process
+
+    temp_dir_name = os.environ.get('TEMP_DIR')
+    destinationFile = destinationFile.replace(temp_dir_name, os.environ.get('EMSCRIPTEN_TEMP_FILES_DIR'))
+
+    if not os.path.exists(os.path.dirname(destinationFile)):
+      os.makedirs(os.path.dirname(destinationFile))
+
+    print >> sys.stderr, "Copying {} to {}".format(file_to_process, destinationFile)
+    shutil.copyfile(file_to_process, destinationFile)
+
+def eliminate_duplicate_funcs(file_name):
+  num_passes = 5
+
+  if os.environ.get('EMCC_DUMP_EQUIVALENT_FUNCTIONS'):
+    # Remove previous log file if it exists
+    equivalent_fn_json_file = file_name + ".equivalent_functions.json"
+    if os.path.isfile(equivalent_fn_json_file):
+      print >> sys.stderr, "Deleting old json: " + equivalent_fn_json_file
+      os.remove(equivalent_fn_json_file)
+
+  dupe_func_passes_str = shared.Settings.ELIMINATE_DUPLICATE_FUNCTIONS_PASSES
+  if dupe_func_passes_str:
+    num_passes = int(dupe_func_passes_str)
```

Comment:
Good catch - this is fallout from an earlier suggestion to use Settings instead of the EMCC_DUPLICATE_FUNCTION_ELIMINATION_NUM_PASSES variable.